distribution/tools/entitlement-runtime/src/main/java/org/elasticsearch/entitlement/runtime/checks/EntitlementChecksService.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.entitlement.runtime.checks;
+
+import org.elasticsearch.entitlement.checks.EntitlementChecks;
+import org.elasticsearch.entitlement.checks.EntitlementProvider;
+import org.elasticsearch.entitlement.runtime.api.Entitlement;
+import org.elasticsearch.entitlement.runtime.api.FlagEntitlement;
+import org.elasticsearch.entitlement.runtime.api.NotEntitledException;
+
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static java.util.Collections.newSetFromMap;
+import static org.elasticsearch.entitlement.runtime.internals.EntitlementInternals.isActive;
+import static org.elasticsearch.entitlement.runtime.internals.EntitlementInternals.isFrozen;
+
+/**
+ * Implementation of the {@link EntitlementChecks} interface.
+ * The trampoline module loads this object via SPI.
+ */
+public class EntitlementChecksService implements EntitlementChecks {
+    /**
+     * @return the same instance of {@link EntitlementChecksService} returned by {@link EntitlementProvider}.
+     */
+    public static EntitlementChecksService get() {
+        return (EntitlementChecksService) EntitlementProvider.checks();
+    }
+
+    private final Set<Module> entitledToExit = newSetFromMap(new ConcurrentHashMap<>());
+
+    public void grant(Module module, Entitlement e) {
+        if (isFrozen) {
+            throw new IllegalStateException("Entitlement grants are frozen");
+        }
+        switch (e) {
+            case FlagEntitlement f -> {
+                switch (f) {
+                    case EXIT_JVM -> entitledToExit.add(module);
+                }
+            }
+        }
+    }
+
+    /**
+     * Mainly for testing purposes.
+     */
+    public void revokeAll() {
+        if (isFrozen) {
+            throw new IllegalStateException("Entitlement grants are frozen");
+        }
+        entitledToExit.clear();
+    }
+
+    /**
+     * Causes entitlements to be enforced.
+     */
+    public void activate() {
+        isActive = true;
+    }
+
+    /**
+     * Disallows changes to the entitlement grants.
+     */
+    public void freeze() {
+        isFrozen = true;
+    }
+
+    @Override
+    public void checkSystemExit(Class<?> callerClass, System system, int status) {
+        checkFlagEntitlement(callerClass, entitledToExit);
```

Comment:
Let's not assume any policy structure, that is being worked on separately. Let's instead use a hardcoded check for server here:

```
if (callerClass.getModule().equals(Elasticsearch.class.getModule()) == false) {
    // throw
}
```

Reply:
Jack is working on how to describe the needs of the plugins; this is describing the needs of the library methods.

In any case, these classes are so simple they can be treated as placeholders to be iterated upon in the coming days.